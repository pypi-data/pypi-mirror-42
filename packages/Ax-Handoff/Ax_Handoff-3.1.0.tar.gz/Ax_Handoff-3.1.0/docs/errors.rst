==============================================================================
Errors & Exceptions
==============================================================================

The ``axonchisel.handoff.error`` module contains all Exceptions raised by 
this package:
    
- ``SerializeError``:
    Error encoding an object to wire format.
        
- ``UnserializeError``
    Error decoding an object from wire format.

- ``DataTamperedError``
    Subclass of UnserializeError indicating data
    was tampered with or corrupted in transit.

