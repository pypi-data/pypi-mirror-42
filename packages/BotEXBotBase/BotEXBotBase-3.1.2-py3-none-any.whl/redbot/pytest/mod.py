import pytest

__all__ = ["mod"]


@pytest.fixture
def mod(config, monkeypatch):
    from redbot.core import Config

    with monkeypatch.context() as m:
        m.setattr(Config, "get_conf", lambda *args, **kwargs: config)
        from redbot.core import modlog

        modlog._register_defaults()
        return modlog
