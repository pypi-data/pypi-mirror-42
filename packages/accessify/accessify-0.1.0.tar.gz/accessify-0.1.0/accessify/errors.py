"""
Provide errors for accessibility levels.
"""
INACCESSIBLE_DUE_TO_ITS_PROTECTION_LEVEL_EXCEPTION_MESSAGE = \
    '{class_name}.{method_name}() is inaccessible due to its protection level'


class InaccessibleDueToItsProtectionLevelException(Exception):
    """
    Inaccessible due to its protection level exception.
    """

    def __init__(self, message):
        self.message = message