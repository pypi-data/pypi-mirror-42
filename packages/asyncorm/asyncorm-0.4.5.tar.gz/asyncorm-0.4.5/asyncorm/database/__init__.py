from asyncorm.database.db_manager import PostgresManager
from asyncorm.database.db_cursor import Cursor

__all__ = ["PostgresManager", "Cursor"]
