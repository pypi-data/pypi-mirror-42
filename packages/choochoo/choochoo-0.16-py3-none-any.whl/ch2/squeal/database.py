
from contextlib import contextmanager

from sqlalchemy import create_engine, event
from sqlalchemy.engine import Engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.sql.functions import count

from . import *
from .support import Base
from ..command.args import DATABASE, NamespaceWithVariables, NO_OP, parser
from ..lib.log import make_log

# mention these so they are "created"
Source,  Interval,
ActivityGroup, ActivityJournal, ActivityTimespan,
Topic, TopicJournal,
StatisticName, StatisticJournal, StatisticJournalInteger, StatisticJournalFloat, StatisticJournalText,
Segment, SegmentJournal
Pipeline
MonitorJournal
Constant, SystemConstant
ActivitySimilarity, ActivityNearby


# https://stackoverflow.com/questions/13712381/how-to-turn-on-pragma-foreign-keys-on-in-sqlalchemy-migration-script-or-conf
@event.listens_for(Engine, "connect")
def fk_pragma_on_connect(dbapi_con, _con_record):
    cursor = dbapi_con.cursor()
    cursor.execute("PRAGMA foreign_keys=ON;")  # https://www.sqlite.org/pragma.html#pragma_foreign_keys
    cursor.execute("PRAGMA temp_store=MEMORY;")  # https://www.sqlite.org/pragma.html#pragma_temp_store
    cursor.execute("PRAGMA threads=4;")  # https://www.sqlite.org/pragma.html#pragma_threads
    cursor.execute("PRAGMA cache_size=-1000000;")  # 1GB  https://www.sqlite.org/pragma.html#pragma_cache_size
    cursor.execute("PRAGMA secure_delete=OFF;")  # https://www.sqlite.org/pragma.html#pragma_secure_delete
    # cursor.execute("PRAGMA journal_mode=WAL;")  # https://www.sqlite.org/wal.html
    cursor.close()


@event.listens_for(Engine, 'close')
def analyxe_pragma_on_close(dbapi_con, _con_record):
    cursor = dbapi_con.cursor()
    cursor.execute("PRAGMA optimize;")  # https://www.sqlite.org/pragma.html#pragma_optimize
    cursor.close()


class Database:

    def __init__(self, args, log):
        self._log = log
        path = args.file(DATABASE)
        self._log.info('Using database at %s' % path)
        self.engine = create_engine('sqlite:///%s' % path, echo=False)
        self.session = sessionmaker(bind=self.engine)
        self.__create_tables()

    def __create_tables(self):
        if self.is_empty(tables=True):
            self._log.info('Creating tables')
            Base.metadata.create_all(self.engine)

    @contextmanager
    def session_context(self):
        session = self.session()
        try:
            yield session
            session.commit()
        except:
            session.rollback()
            raise
        finally:
            session.close()

    def is_empty(self, tables=False):
        if tables:
            # https://stackoverflow.com/questions/33053241/sqlalchemy-if-table-does-not-exist
            return not self.engine.dialect.has_table(self.engine, Source.__tablename__)
        else:
            with self.session_context() as s:
                n_topics = s.query(count(Topic.id)).scalar()
                n_activities = s.query(count(ActivityGroup.id)).scalar()
                n_statistics = s.query(count(StatisticName.id)).scalar()
            return not (n_topics + n_activities + n_statistics)


def add(s, instance):
    s.add(instance)
    return instance


def connect(args):
    '''
    Bootstrap from commandline-like args.
    '''
    p = parser()
    if len(args) == 1:
        args = args[0].split()
    elif args:
        args = list(args)
    else:
        args = []
    args.append(NO_OP)
    ns = NamespaceWithVariables(p.parse_args(args))
    log = make_log(ns)
    db = Database(ns, log)
    return ns, log, db

