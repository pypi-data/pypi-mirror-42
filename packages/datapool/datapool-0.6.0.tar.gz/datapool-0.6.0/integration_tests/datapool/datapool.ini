[db]
connection_string = sqlite+pysqlite:////Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz/.simple.db

[backup_landing_zone]
folder = 

[r]
extension = .r
executable = /usr/local/bin/R

[matlab]
extension = .m
executable = /Applications/MATLAB_R2016a.app/bin/matlab

[julia]
extension = .jl
version = 0.5.2
executable = /usr/local/bin/julia

[python]
extension = .py
executable = /Users/uweschmitt/Projects/eawag_datapool/venv/bin/python3

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[landing_zone]
folder = /Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz

