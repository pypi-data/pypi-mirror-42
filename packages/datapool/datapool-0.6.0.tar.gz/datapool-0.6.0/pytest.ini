[pytest]
addopts = -rs --full-trace
testpaths = tests
