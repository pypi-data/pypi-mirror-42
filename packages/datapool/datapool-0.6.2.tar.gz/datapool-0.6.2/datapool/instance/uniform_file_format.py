# encoding: utf-8
from __future__ import print_function, division, absolute_import

import csv
import datetime
import io
import os

from datapool.utils import iter_to_list


def _check_header(header):

    header = list(header)

    if "site" in header:
        header.remove("site")
    else:
        for n in "xyz":
            if n not in header:
                raise ValueError(
                    "header {!r} must have either 'site' or 'x', 'y' and 'z'".format(
                        header
                    )
                )
            header.remove(n)

    header = set(header)
    fix_column_names = set(("timestamp", "parameter", "value"))
    if header != fix_column_names:
        missing = ["'{}'".format(name) for name in fix_column_names - header]
        invalid = ["'{}'".format(name) for name in header - fix_column_names]
        msg = ""
        if missing:
            msg += "missing: {}".format(", ".join(sorted(missing)))
        if invalid:
            msg += "invalid: {}".format(", ".join(sorted(invalid)))
        raise ValueError("invalid header ({})".format(msg))


@iter_to_list
def read_from_file(path):
    if not os.path.exists(path):
        raise ValueError("{} does not exist".format(path))

    with open(path, "r", encoding="ascii") as fh:
        yield from _read_from_fh(fh)


@iter_to_list
def read_from_string(data):
    fh = io.StringIO(data)
    yield from _read_from_fh(fh)


def can_be_converted_to_float(value):
    try:
        value = float(value)
    except ValueError:
        return False
    return True


def parse_timestamp(timestamp):
    return datetime.datetime.strptime(timestamp, "%Y-%m-%d %H:%M:%S")


def _read_from_fh(fh):
    reader = csv.reader(fh, delimiter=";")
    header = [f.strip() for f in next(reader)]
    _check_header(header)
    for i, row in enumerate(reader):
        if len(row) < len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is incomplete".format(i + 1, row))
        if len(row) > len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is to long".format(i + 1, row))

        row = (cell.strip() for cell in row)
        row_dict = dict(zip(header, row))
        yield row_dict


@iter_to_list
def check_rows(row_dicts):

    if not row_dicts:
        return

    for i, row_dict in enumerate(row_dicts):

        value = row_dict["value"]
        if not can_be_converted_to_float(value):
            yield "row {}: value '{}' is not a float".format(i, value)
        else:
            row_dict["value"] = float(value)

        timestamp = row_dict["timestamp"]
        try:
            row_dict["timestamp"] = parse_timestamp(timestamp)
        except ValueError:
            yield "row {}: timstamp '{}' either has wrong format or is invalid".format(
                i, timestamp
            )

        has_xyz = all(name in row_dict and row_dict[name] is not None for name in "xyz")
        not_empty = ("parameter",)
        if not has_xyz:
            not_empty += ("site",)

        for name in not_empty:
            if row_dict[name].strip() == "":
                yield "row {}: field {} is empty".format(i, name)


@iter_to_list
def to_signals(row_dicts):

    for row in row_dicts:
        for f in "xyz":
            if f in row:
                row["coord_" + f] = row[f]
                del row[f]

    from .domain_objects import Signal

    return list(map(Signal, row_dicts))
