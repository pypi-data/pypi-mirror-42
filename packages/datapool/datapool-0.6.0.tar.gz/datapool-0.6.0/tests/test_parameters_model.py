# encoding: utf-8
from __future__ import print_function, division, absolute_import

from functools import partial

import pytest

from datapool.instance.yaml_parsers import parse_parameters
from datapool.database import setup_db, dump_db
from datapool.errors import ConsistencyError
from datapool.instance.parameters_model import check_and_commit
from datapool.instance.pretty_printers import pretty_print

from .yaml_helpers import load

load_parameters = partial(load, parse_parameters)


def test_load_valid_parameters(data_path, regtest, regtest_print, regtest_logger, lz):
    with regtest_logger():
        excs, all_parameters = load_parameters(lz)
    assert not excs, str(excs)
    assert len(all_parameters) == 1
    parameters = all_parameters[0]

    regtest_print("parameter objects read from yaml:")
    for parameter in parameters:
        pretty_print(parameter, file=regtest)


def test_load_empty_parameters(data_path, regtest, regtest_print, regtest_logger, lz):
    lz.overwrite_file("data/parameters.yaml", data_path("yamls/empty.yaml"))

    with regtest_logger():
        excs, __ = load_parameters(lz)

    assert excs
    for exc in excs:
        regtest_print(exc)


def test_load_invalid_parameters(data_path, regtest, regtest_print, regtest_logger, lz):
    lz.overwrite_file(
        "data/parameters.yaml", data_path("yamls/parameters_invalid.yaml")
    )

    with regtest_logger():
        excs, __ = load_parameters(lz)

    assert excs
    for exc in excs:
        regtest_print(exc)


@pytest.fixture(scope="function")
def setup(data_path, regtest_logger, regtest, regtest_print, config, lz):
    with regtest_logger():
        engine = setup_db(config.db)

    with regtest_logger():
        excs, all_parameters = load_parameters(lz)

    assert not excs, str(excs)
    assert len(all_parameters) == 1
    parameters = all_parameters[0]

    regtest_print("parameters object read from yaml:")
    pretty_print(parameters, file=regtest)

    with regtest_logger():
        return config, engine, check_and_commit(parameters, engine)


def test_commit_valid_parameters(setup, regtest_print, regtest_logger, regtest):
    config, engine, parameter_dbos = setup
    regtest_print()
    for parameter_dbo in parameter_dbos:
        regtest_print("site row:", parameter_dbo)
    regtest_print()
    with regtest_logger():
        dump_db(config.db, file=regtest)


def _inject_and_test(
    engine,
    yaml_file,
    regtest_logger,
    lz,
    regtest_print,
    regtest,
    config,
    expect_exception=True,
):
    lz.overwrite_file("data/parameters.yaml", yaml_file)

    with regtest_logger():
        excs, all_parameters = load_parameters(lz)

    assert not excs, str(excs)
    assert len(all_parameters) == 1
    parameters = all_parameters[0]

    regtest_print("parameters object read from yaml:")
    for parameter in parameters:
        pretty_print(parameter, file=regtest)

    if expect_exception:
        with pytest.raises(ConsistencyError) as e:
            with regtest_logger():
                for dbo in check_and_commit(parameters, engine):
                    print(dbo, file=regtest)
        print(file=regtest)
        print(e.value, file=regtest)
        print(file=regtest)
    else:
        with regtest_logger():
            for dbo in check_and_commit(parameters, engine):
                print(dbo, file=regtest)

    dump_db(config.db, file=regtest)


def test_commit_modified_parameters(
    setup, data_path, regtest_logger, regtest, regtest_print, config, lz
):
    config, engine, __ = setup
    _inject_and_test(
        engine,
        data_path("yamls/parameters_modified.yaml"),
        regtest_logger,
        lz,
        regtest_print,
        regtest,
        config,
    )


def test_commit_modified_parameters_2(
    setup, data_path, regtest_logger, regtest, regtest_print, config, lz
):
    # check allowed updates
    config, engine, __ = setup
    _inject_and_test(
        engine,
        data_path("yamls/parameters_modified_2.yaml"),
        regtest_logger,
        lz,
        regtest_print,
        regtest,
        config,
        expect_exception=False,
    )


def test_commit_shortened_parameters(
    setup, data_path, regtest_logger, regtest, regtest_print, config, lz
):
    config, engine, __ = setup
    _inject_and_test(
        engine,
        data_path("yamls/parameters_shortened.yaml"),
        regtest_logger,
        lz,
        regtest_print,
        regtest,
        config,
    )
