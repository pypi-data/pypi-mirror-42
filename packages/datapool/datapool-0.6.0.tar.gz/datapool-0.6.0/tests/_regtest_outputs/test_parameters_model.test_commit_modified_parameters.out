INFO     eawag.datapool  connected to db sqlite+pysqlite:///<tmpdir_from_fixture>/datapool.db
INFO     eawag.datapool  connected to db sqlite+pysqlite:///<tmpdir_from_fixture>/datapool.db
INFO     eawag.datapool  created all tables of db sqlite+pysqlite:///<tmpdir_from_fixture>/datapool.db
INFO     eawag.datapool  use parse_parameters to parse all files from <tmpdir_from_fixture>/landing_zone
INFO     eawag.datapool  load <tmpdir_from_fixture>/landing_zone/data/parameters.yaml
INFO     eawag.datapool  number of parameters found: 14
parameters object read from yaml:
Parameters:
    name        : 'Absorbance 200.00_nm'
    description : 'absorbance at 200.00 nm wavelength'
    unit        : 'm-1'

    name        : 'Absorbance 202.50_nm'
    description : 'absorbance at 202.50 nm wavelength'
    unit        : 'm-1'

    name        : 'rain intensity'
    description : None
    unit        : 'mm/h'

    name        : 'Water Level'
    description : None
    unit        : 'm'

    name        : 'Average Flow Velocity'
    description : None
    unit        : 'm/s'

    name        : 'Flow'
    description : None
    unit        : 'm/s'

    name        : 'Temperature'
    description : None
    unit        : 'C'

    name        : 'Surface Flow Velocity'
    description : None
    unit        : 'm/s'

    name        : 'Distance'
    description : None
    unit        : 'km'

    name        : 'Distance Reading Count'
    description : None
    unit        : '1'

    name        : 'Surcharge Level'
    description : None
    unit        : 'm'

    name        : 'Peak to Mean Ratio'
    description : None
    unit        : '1'

    name        : 'Battery Voltage'
    description : None
    unit        : 'V'

    name        : 'Number of Samples'
    description : None
    unit        : '1'
INFO     eawag.datapool  called _commit for parameter Absorbance 200.00_nm
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Absorbance 200.00_nm, unit=m-1, description=absorbance at 200.00 nm wavelength>
INFO     eawag.datapool  called _commit for parameter Absorbance 202.50_nm
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Absorbance 202.50_nm, unit=m-1, description=absorbance at 202.50 nm wavelength>
INFO     eawag.datapool  called _commit for parameter Average Flow Velocity
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Average Flow Velocity, unit=m/s, description=None>
INFO     eawag.datapool  called _commit for parameter Battery Voltage
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Battery Voltage, unit=V, description=None>
INFO     eawag.datapool  called _commit for parameter Distance
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Distance, unit=km, description=None>
INFO     eawag.datapool  called _commit for parameter Distance Reading Count
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Distance Reading Count, unit=1, description=None>
INFO     eawag.datapool  called _commit for parameter Flow
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Flow, unit=m/s, description=None>
INFO     eawag.datapool  called _commit for parameter Number of Samples
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Number of Samples, unit=1, description=None>
INFO     eawag.datapool  called _commit for parameter Peak to Mean Ratio
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Peak to Mean Ratio, unit=1, description=None>
INFO     eawag.datapool  called _commit for parameter Surcharge Level
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Surcharge Level, unit=m, description=None>
INFO     eawag.datapool  called _commit for parameter Surface Flow Velocity
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Surface Flow Velocity, unit=m/s, description=None>
INFO     eawag.datapool  called _commit for parameter Temperature
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Temperature, unit=C, description=None>
INFO     eawag.datapool  called _commit for parameter Water Level
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Water Level, unit=m, description=None>
INFO     eawag.datapool  called _commit for parameter rain intensity
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=rain intensity, unit=mm/h, description=None>
INFO     eawag.datapool  use parse_parameters to parse all files from <tmpdir_from_fixture>/landing_zone
INFO     eawag.datapool  load <tmpdir_from_fixture>/landing_zone/data/parameters.yaml
INFO     eawag.datapool  number of parameters found: 3
parameters object read from yaml:
Parameter:
    name        : 'Absorbance 200.00_nm'
    description : 'absorbance at 200.00 nm wavelength'
    unit        : 'm-17'
Parameter:
    name        : 'Absorbance 202.50_nm'
    description : 'absorbance at 202.50 nm wavelength'
    unit        : 'm-1'
Parameter:
    name        : 'rain intensity'
    description : 'old description was empty'
    unit        : 'mm/h'
ERROR    eawag.datapool  detected removed parameters:
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Average Flow Velocity'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : 'm/s'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Battery Voltage'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : 'V'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Distance'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : 'km'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Distance Reading Count'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : '1'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Flow'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : 'm/s'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Number of Samples'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : '1'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Peak to Mean Ratio'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : '1'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Surcharge Level'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : 'm'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Surface Flow Velocity'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : 'm/s'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Temperature'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : 'C'
ERROR    eawag.datapool  ParameterDbo:
ERROR    eawag.datapool      name        : 'Water Level'
ERROR    eawag.datapool      description : None
ERROR    eawag.datapool      unit        : 'm'

parameter Average Flow Velocity was removed

INFO     eawag.datapool  connected to db sqlite+pysqlite:///<tmpdir_from_fixture>/datapool.db
table comment:

   comment_id     text     timestamp     author  
   ----------     ----     ---------     ------  

table parameter:

   parameter_id     name                       description                            unit  
   ------------     ----------------------     ----------------------------------     ----  
   1                Absorbance 200.00_nm       absorbance at 200.00 nm wavelength     m-1   
   2                Absorbance 202.50_nm       absorbance at 202.50 nm wavelength     m-1   
   3                Average Flow Velocity      None                                   m/s   
   4                Battery Voltage            None                                   V     
   5                Distance                   None                                   km    
   6                Distance Reading Count     None                                   1     
   7                Flow                       None                                   m/s   
   8                Number of Samples          None                                   1     
   9                Peak to Mean Ratio         None                                   1     
   10               Surcharge Level            None                                   m     
   11               Surface Flow Velocity      None                                   m/s   
   12               Temperature                None                                   C     
   13               Water Level                None                                   m     
   14               rain intensity             None                                   mm/h  

table parameter_averaging:

   parameter_averaging_id     parameter_id     source_id     integration_length_x     integration_length_y     integration_angle     integration_time  
   ----------------------     ------------     ---------     --------------------     --------------------     -----------------     ----------------  

table picture:

   picture_id     site_id     filename     description     timestamp     data  
   ----------     -------     --------     -----------     ---------     ----  

table project:

   project_id     title     description  
   ----------     -----     -----------  

table quality:

   quality_id     flag     method  
   ----------     ----     ------  

table signal:

   signal_id     value     timestamp     parameter_id     source_id     site_id     coord_x     coord_y     coord_z  
   ---------     -----     ---------     ------------     ---------     -------     -------     -------     -------  

table signal_quality:

   signal_quality_id     quality_id     timestamp     author  
   -----------------     ----------     ---------     ------  

table signals_comments_association:

   signal_id     comment_id  
   ---------     ----------  

table signals_signal_quality_association:

   signal_id     signal_quality_id  
   ---------     -----------------  

table site:

   site_id     name     description     street     postcode     city     coord_x     coord_y     coord_z  
   -------     ----     -----------     ------     --------     ----     -------     -------     -------  

table source:

   source_id     source_type_id     name     description     serial     manufacturer     manufacturing_date  
   ---------     --------------     ----     -----------     ------     ------------     ------------------  

table source_type:

   source_type_id     name     description  
   --------------     ----     -----------  

table special_value_definition:

   special_value_definition_id     source_type_id     description     categorical_value     numerical_value  
   ---------------------------     --------------     -----------     -----------------     ---------------  

